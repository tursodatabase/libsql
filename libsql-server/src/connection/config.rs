use parking_lot::Mutex;
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::{fs, io};
use url::Url;

use crate::error::Error;
use crate::{Result, LIBSQL_PAGE_SIZE};

#[derive(Debug)]
pub struct DatabaseConfigStore {
    config_path: PathBuf,
    tmp_config_path: PathBuf,
    config: Mutex<Arc<DatabaseConfig>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseConfig {
    #[serde(default)]
    pub block_reads: bool,
    #[serde(default)]
    pub block_writes: bool,
    /// The reason why operations are blocked. This will be included in [`Error::Blocked`].
    #[serde(default)]
    pub block_reason: Option<String>,
    /// maximum db size (in pages)
    #[serde(default = "default_max_size")]
    pub max_db_pages: u64,
    #[serde(default)]
    pub heartbeat_url: Option<Url>,
    #[serde(default)]
    pub bottomless_db_id: Option<String>,
}

const fn default_max_size() -> u64 {
    bytesize::ByteSize::pb(1000).as_u64() / LIBSQL_PAGE_SIZE
}

impl Default for DatabaseConfig {
    fn default() -> Self {
        Self {
            block_reads: Default::default(),
            block_writes: Default::default(),
            block_reason: Default::default(),
            max_db_pages: default_max_size(),
            heartbeat_url: None,
            bottomless_db_id: None,
        }
    }
}

impl DatabaseConfigStore {
    pub fn load(db_path: &Path) -> Result<Self> {
        let config_path = db_path.join("config.json");
        let tmp_config_path = db_path.join("config.json~");

        let config = match fs::read(&config_path) {
            Ok(data) => serde_json::from_slice(&data)?,
            Err(err) if err.kind() == io::ErrorKind::NotFound => DatabaseConfig::default(),
            Err(err) => return Err(Error::IOError(err)),
        };

        Ok(Self {
            config_path,
            tmp_config_path,
            config: Mutex::new(Arc::new(config)),
        })
    }

    #[cfg(test)]
    pub fn new_test() -> Self {
        Self {
            config_path: "".into(),
            tmp_config_path: "".into(),
            config: Mutex::new(Arc::new(DatabaseConfig::default())),
        }
    }

    pub fn get(&self) -> Arc<DatabaseConfig> {
        self.config.lock().clone()
    }

    pub fn store(&self, config: DatabaseConfig) -> Result<()> {
        let data = serde_json::to_vec_pretty(&config)?;
        fs::write(&self.tmp_config_path, data)?;
        fs::rename(&self.tmp_config_path, &self.config_path)?;
        *self.config.lock() = Arc::new(config);
        Ok(())
    }
}
