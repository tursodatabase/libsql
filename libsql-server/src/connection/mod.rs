use libsql_replication::rpc::replication::NAMESPACE_METADATA_KEY;
use std::ops::Deref;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;
use tokio::time::{Duration, Instant};

use futures::Future;
use tokio::{sync::Semaphore, time::timeout};
use tonic::metadata::BinaryMetadataValue;

use crate::auth::Authenticated;
use crate::error::Error;
use crate::metrics::{
    CONCURRENT_CONNECTIONS_COUNT, CONNECTION_ALIVE_DURATION, CONNECTION_CREATE_TIME,
};
use crate::namespace::meta_store::MetaStore;
use crate::namespace::NamespaceName;
use crate::query::{Params, Query};
use crate::query_analysis::Statement;
use crate::query_result_builder::{IgnoreResult, QueryResultBuilder};
use crate::replication::FrameNo;
use crate::Result;

use self::program::{Cond, DescribeResponse, Program, Step};

pub mod config;
pub mod connection_manager;
pub mod dump;
pub mod libsql;
pub mod program;
pub mod write_proxy;

#[cfg(not(test))]
const TXN_TIMEOUT: Duration = Duration::from_secs(5);
#[cfg(test)]
const TXN_TIMEOUT: Duration = Duration::from_millis(100);

#[derive(Clone)]
pub struct RequestContext {
    /// Authentication for this request
    auth: Authenticated,
    /// current namespace
    namespace: NamespaceName,
    meta_store: MetaStore,
}

impl RequestContext {
    pub fn new(auth: Authenticated, namespace: NamespaceName, meta_store: MetaStore) -> Self {
        Self {
            auth,
            namespace,
            meta_store,
        }
    }

    pub fn upgrade_grpc_request<T>(&self, req: &mut tonic::Request<T>) {
        let namespace = BinaryMetadataValue::from_bytes(self.namespace.as_slice());
        req.metadata_mut()
            .insert_bin(NAMESPACE_METADATA_KEY, namespace);
        self.auth.upgrade_grpc_request(req);
    }

    pub fn namespace(&self) -> &NamespaceName {
        &self.namespace
    }

    pub fn auth(&self) -> &Authenticated {
        &self.auth
    }
}

#[async_trait::async_trait]
pub trait Connection: Send + Sync + 'static {
    /// Executes a query program
    async fn execute_program<B: QueryResultBuilder>(
        &self,
        pgm: Program,
        ctx: RequestContext,
        response_builder: B,
        replication_index: Option<FrameNo>,
    ) -> Result<B>;

    /// Execute all the queries in the batch sequentially.
    /// If an query in the batch fails, the remaining queries are ignores, and the batch current
    /// transaction (if any) is rolledback.
    async fn execute_batch_or_rollback<B: QueryResultBuilder>(
        &self,
        batch: Vec<Query>,
        ctx: RequestContext,
        result_builder: B,
        replication_index: Option<FrameNo>,
    ) -> Result<B> {
        let batch_len = batch.len();
        let mut steps = make_batch_program(batch);

        if !steps.is_empty() {
            // We add a conditional rollback step if the last step was not successful.
            steps.push(Step {
                query: Query {
                    stmt: Statement::parse("ROLLBACK").next().unwrap().unwrap(),
                    params: Params::empty(),
                    want_rows: false,
                },
                cond: Some(Cond::Not {
                    cond: Box::new(Cond::Ok {
                        step: steps.len() - 1,
                    }),
                }),
            })
        }

        let pgm = Program::new(steps);

        // ignore the rollback result
        let builder = result_builder.take(batch_len);
        let builder = self
            .execute_program(pgm, ctx, builder, replication_index)
            .await?;

        Ok(builder.into_inner())
    }

    /// Execute all the queries in the batch sequentially.
    /// If an query in the batch fails, the remaining queries are ignored
    async fn execute_batch<B: QueryResultBuilder>(
        &self,
        batch: Vec<Query>,
        ctx: RequestContext,
        result_builder: B,
        replication_index: Option<FrameNo>,
    ) -> Result<B> {
        let steps = make_batch_program(batch);
        let pgm = Program::new(steps);
        self.execute_program(pgm, ctx, result_builder, replication_index)
            .await
    }

    async fn rollback(&self, ctx: RequestContext) -> Result<()> {
        self.execute_batch(
            vec![Query {
                stmt: Statement::parse("ROLLBACK").next().unwrap().unwrap(),
                params: Params::empty(),
                want_rows: false,
            }],
            ctx,
            IgnoreResult,
            None,
        )
        .await?;

        Ok(())
    }

    /// Parse the SQL statement and return information about it.
    async fn describe(
        &self,
        sql: String,
        ctx: RequestContext,
        replication_index: Option<FrameNo>,
    ) -> Result<Result<DescribeResponse>>;

    /// Check whether the connection is in autocommit mode.
    async fn is_autocommit(&self) -> Result<bool>;

    /// Calls for database checkpoint (if supported).
    async fn checkpoint(&self) -> Result<()>;

    // Calls for database vacuum (if supported).
    async fn vacuum_if_needed(&self) -> Result<()>;

    fn diagnostics(&self) -> String;
}

fn make_batch_program(batch: Vec<Query>) -> Vec<Step> {
    let mut steps = Vec::with_capacity(batch.len());
    for (i, query) in batch.into_iter().enumerate() {
        let cond = if i > 0 {
            // only execute if the previous step was a success
            Some(Cond::Ok { step: i - 1 })
        } else {
            None
        };

        let step = Step { cond, query };
        steps.push(step);
    }
    steps
}

#[async_trait::async_trait]
pub trait MakeConnection: Send + Sync + 'static {
    type Connection: Connection;

    /// Create a new connection of type Self::Connection
    async fn create(&self) -> Result<Self::Connection, Error>;

    fn throttled(
        self,
        semaphore: Arc<Semaphore>,
        timeout: Option<Duration>,
        max_total_response_size: u64,
        max_concurrent_requests: u64,
    ) -> MakeThrottledConnection<Self>
    where
        Self: Sized,
    {
        MakeThrottledConnection::new(
            semaphore,
            self,
            timeout,
            max_total_response_size,
            max_concurrent_requests,
        )
    }

    fn map<F, T>(self, f: F) -> Map<Self, F>
    where
        F: Fn(Self::Connection) -> T + Send + Sync + 'static,
        Self: Sized,
    {
        Map { inner: self, f }
    }
}

pub struct Map<T, F> {
    inner: T,
    f: F,
}

#[async_trait::async_trait]
impl<F, T, O> MakeConnection for Map<T, F>
where
    F: Fn(T::Connection) -> O + Send + Sync + 'static,
    T: MakeConnection,
    O: Connection,
{
    type Connection = O;

    async fn create(&self) -> Result<Self::Connection, Error> {
        let conn = self.inner.create().await?;
        Ok((self.f)(conn))
    }
}

#[async_trait::async_trait]
impl<T: MakeConnection> MakeConnection for Arc<T> {
    type Connection = T::Connection;

    async fn create(&self) -> Result<Self::Connection, Error> {
        self.as_ref().create().await
    }
}

#[async_trait::async_trait]
impl<F, C, Fut> MakeConnection for F
where
    F: Fn() -> Fut + Send + Sync + 'static,
    Fut: Future<Output = Result<C, Error>> + Send,
    C: Connection + Sync + Send + 'static,
{
    type Connection = C;

    async fn create(&self) -> Result<Self::Connection, Error> {
        let db = (self)().await?;
        Ok(db)
    }
}

pub struct MakeThrottledConnection<F> {
    semaphore: Arc<Semaphore>,
    connection_maker: F,
    timeout: Option<Duration>,
    // Max memory available for responses. High memory pressure
    // will result in reducing concurrency to prevent out-of-memory errors.
    max_total_response_size: u64,
    waiters: AtomicUsize,
    max_concurrent_requests: u64,
}

impl<F> MakeThrottledConnection<F> {
    fn new(
        semaphore: Arc<Semaphore>,
        connection_maker: F,
        timeout: Option<Duration>,
        max_total_response_size: u64,
        max_concurrent_requests: u64,
    ) -> Self {
        Self {
            semaphore,
            connection_maker,
            timeout,
            max_total_response_size,
            waiters: AtomicUsize::new(0),
            max_concurrent_requests,
        }
    }

    // How many units should be acquired from the semaphore,
    // depending on current memory pressure.
    fn units_to_take(&self) -> u32 {
        let total_response_size = crate::query_result_builder::TOTAL_RESPONSE_SIZE
            .load(std::sync::atomic::Ordering::Relaxed) as u64;
        if total_response_size * 2 > self.max_total_response_size {
            tracing::trace!("High memory pressure, reducing concurrency");
            16
        } else if total_response_size * 4 > self.max_total_response_size {
            tracing::trace!("Medium memory pressure, reducing concurrency");
            4
        } else {
            1
        }
    }
}

struct WaitersGuard<'a> {
    pub waiters: &'a AtomicUsize,
}

impl<'a> WaitersGuard<'a> {
    fn new(waiters: &'a AtomicUsize) -> Self {
        waiters.fetch_add(1, Ordering::Relaxed);
        Self { waiters }
    }
}

impl Drop for WaitersGuard<'_> {
    fn drop(&mut self) {
        self.waiters.fetch_sub(1, Ordering::Relaxed);
    }
}

#[async_trait::async_trait]
impl<F: MakeConnection> MakeConnection for MakeThrottledConnection<F> {
    type Connection = TrackedConnection<F::Connection>;

    async fn create(&self) -> Result<Self::Connection, Error> {
        let before_create = Instant::now();
        // If the memory pressure is high, request more units to reduce concurrency.
        tracing::trace!(
            "Available semaphore units: {}",
            self.semaphore.available_permits()
        );
        let units = self.units_to_take();
        let waiters_guard = WaitersGuard::new(&self.waiters);
        if (waiters_guard.waiters.load(Ordering::Relaxed) as u64) >= self.max_concurrent_requests {
            return Err(Error::TooManyRequests);
        }
        let fut = self.semaphore.clone().acquire_many_owned(units);
        let mut permit = match self.timeout {
            Some(t) => timeout(t, fut).await.map_err(|_| Error::DbCreateTimeout)?,
            None => fut.await,
        }
        .expect("semaphore closed");

        let units = self.units_to_take();
        if units > 1 {
            tracing::debug!("Reacquiring {units} units due to high memory pressure");
            let fut = self.semaphore.clone().acquire_many_owned(units);
            let mem_permit = match self.timeout {
                Some(t) => timeout(t, fut).await.map_err(|_| Error::DbCreateTimeout)?,
                None => fut.await,
            }
            .expect("semaphore closed");
            permit.merge(mem_permit);
        }

        let inner = self.connection_maker.create().await?;

        CONCURRENT_CONNECTIONS_COUNT.increment(1.0);
        CONNECTION_CREATE_TIME.record(before_create.elapsed());

        Ok(TrackedConnection {
            permit,
            inner,
            created_at: Instant::now(),
        })
    }
}

#[derive(Debug)]
pub struct TrackedConnection<DB> {
    inner: DB,
    #[allow(dead_code)] // just hold on to it
    permit: tokio::sync::OwnedSemaphorePermit,
    created_at: Instant,
}

impl<T> Drop for TrackedConnection<T> {
    fn drop(&mut self) {
        CONCURRENT_CONNECTIONS_COUNT.decrement(1.0);
        CONNECTION_ALIVE_DURATION.record(self.created_at.elapsed());
    }
}

impl<T> Deref for TrackedConnection<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

#[async_trait::async_trait]
impl<DB: Connection> Connection for TrackedConnection<DB> {
    #[inline]
    async fn execute_program<B: QueryResultBuilder>(
        &self,
        pgm: Program,
        ctx: RequestContext,
        builder: B,
        replication_index: Option<FrameNo>,
    ) -> crate::Result<B> {
        self.inner
            .execute_program(pgm, ctx, builder, replication_index)
            .await
    }

    #[inline]
    async fn describe(
        &self,
        sql: String,
        ctx: RequestContext,
        replication_index: Option<FrameNo>,
    ) -> crate::Result<crate::Result<DescribeResponse>> {
        self.inner.describe(sql, ctx, replication_index).await
    }

    #[inline]
    async fn is_autocommit(&self) -> crate::Result<bool> {
        self.inner.is_autocommit().await
    }

    #[inline]
    async fn checkpoint(&self) -> Result<()> {
        self.inner.checkpoint().await
    }

    #[inline]
    async fn vacuum_if_needed(&self) -> Result<()> {
        self.inner.vacuum_if_needed().await
    }

    #[inline]
    fn diagnostics(&self) -> String {
        self.inner.diagnostics()
    }
}

#[cfg(test)]
pub mod test {
    use super::*;

    #[derive(Debug)]
    struct DummyDb;

    #[async_trait::async_trait]
    impl Connection for DummyDb {
        async fn execute_program<B: QueryResultBuilder>(
            &self,
            _pgm: Program,
            _ctx: RequestContext,
            _builder: B,
            _replication_index: Option<FrameNo>,
        ) -> crate::Result<B> {
            unreachable!()
        }

        async fn describe(
            &self,
            _sql: String,
            _ctx: RequestContext,
            _replication_index: Option<FrameNo>,
        ) -> crate::Result<crate::Result<DescribeResponse>> {
            unreachable!()
        }

        async fn is_autocommit(&self) -> crate::Result<bool> {
            unreachable!()
        }

        async fn checkpoint(&self) -> Result<()> {
            unreachable!()
        }

        async fn vacuum_if_needed(&self) -> Result<()> {
            unreachable!()
        }

        fn diagnostics(&self) -> String {
            "dummy".into()
        }
    }

    #[tokio::test]
    async fn throttle_db_creation() {
        let factory = (|| async { Ok(DummyDb) }).throttled(
            Arc::new(Semaphore::new(10)),
            Some(Duration::from_millis(100)),
            u64::MAX,
            u64::MAX,
        );

        let mut conns = Vec::with_capacity(10);
        for _ in 0..10 {
            conns.push(factory.create().await.unwrap())
        }

        assert!(factory.create().await.is_err());

        drop(conns);

        assert!(factory.create().await.is_ok());
    }
}
