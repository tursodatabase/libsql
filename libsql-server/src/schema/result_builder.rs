use std::time::Duration;

use crate::query_result_builder::QueryResultBuilder;

pub struct SchemaMigrationResultBuilder<B> {
    inner: B,
    step: usize,
    errors: Vec<(usize, String)>,
}

impl<B> SchemaMigrationResultBuilder<B> {
    pub(crate) fn new(inner: B) -> Self {
        Self {
            inner,
            step: 0,
            errors: Vec::new(),
        }
    }

    pub(crate) fn into_inner(self) -> B {
        self.inner
    }

    pub(crate) fn is_success(&self) -> bool {
        self.errors.is_empty()
    }

    pub(crate) fn into_error(self) -> (usize, String) {
        assert!(!self.errors.is_empty());
        self.errors.into_iter().next().unwrap()
    }
}

impl<B: QueryResultBuilder> QueryResultBuilder for SchemaMigrationResultBuilder<B> {
    type Ret = B::Ret;

    fn init(
        &mut self,
        config: &crate::query_result_builder::QueryBuilderConfig,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.step = 0;
        self.inner.init(config)
    }

    fn begin_step(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.step += 1;
        self.inner.begin_step()
    }

    fn finish_step(
        &mut self,
        affected_row_count: u64,
        last_insert_rowid: Option<i64>,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner
            .finish_step(affected_row_count, last_insert_rowid)
    }

    fn step_error(
        &mut self,
        error: crate::error::Error,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.errors.push((self.step, error.to_string()));
        self.inner.step_error(error)
    }

    fn cols_description<'a>(
        &mut self,
        cols: impl IntoIterator<Item = impl Into<crate::query_result_builder::Column<'a>>>,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.cols_description(cols)
    }

    fn begin_rows(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.begin_rows()
    }

    fn begin_row(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.begin_row()
    }

    fn add_row_value(
        &mut self,
        v: rusqlite::types::ValueRef,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.add_row_value(v)
    }

    fn finish_row(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.finish_row()
    }

    fn finish_rows(&mut self) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.finish_rows()
    }

    fn finish(
        &mut self,
        last_frame_no: Option<crate::replication::FrameNo>,
        is_auto_commit: bool,
    ) -> Result<(), crate::query_result_builder::QueryResultBuilderError> {
        self.inner.finish(last_frame_no, is_auto_commit)
    }

    fn into_ret(self) -> Self::Ret {
        self.inner.into_ret()
    }

    fn add_stats(&mut self, rows_read: u64, rows_written: u64, duration: Duration) {
        self.inner.add_stats(rows_read, rows_written, duration);
    }
}
