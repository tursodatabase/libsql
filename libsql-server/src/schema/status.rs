use serde::{Deserialize, Serialize};

/// Represents the status of a migration task
#[derive(Debug, Serialize, Deserialize, Copy, Clone)]
#[repr(u64)]
pub enum MigrationTaskStatus {
    /// The task was enqueued, and shoudl perform a dry run
    Enqueued = 0,
    /// The dry run was successfull
    DryRunSuccess = 1,
    /// The dry run failed
    DryRunFailure = 2,
    /// The migration task should be performed
    Run = 3,
    /// The migration task was a success
    Success = 4,
    /// The migration task was a failure
    Failure = 5,
}

impl MigrationTaskStatus {
    pub fn from_int(i: u64) -> Self {
        match i {
            0 => Self::Enqueued,
            1 => Self::DryRunSuccess,
            2 => Self::DryRunFailure,
            3 => Self::Run,
            4 => Self::Success,
            5 => Self::Failure,
            _ => panic!(),
        }
    }

    pub fn is_failure(&self) -> bool {
        matches!(self, Self::DryRunFailure | Self::Failure)
    }

    pub const fn num_variants() -> usize {
        // the only use of this is to create a compile error if someone adds a variant
        match Self::Enqueued {
            MigrationTaskStatus::Enqueued => (),
            MigrationTaskStatus::DryRunSuccess => (),
            MigrationTaskStatus::DryRunFailure => (),
            MigrationTaskStatus::Run => (),
            MigrationTaskStatus::Success => (),
            MigrationTaskStatus::Failure => (),
        }

        // don't forget to update that!
        6
    }
}

/// Represents the status of a migration job
#[derive(Debug, Serialize, Deserialize, Copy, Clone, PartialEq, Eq)]
#[repr(u64)]
pub enum MigrationJobStatus {
    /// Waiting for all tasks to report the dry run status
    WaitingDryRun = 0,
    /// All tasks reported a successfull dry run status
    DryRunSuccess = 1,
    /// One or more task reported an unsuccessful dry drun
    DryRunFailure = 2,
    /// Waiting for all the tasks to return a successful migration run
    WaitingRun = 3,
    /// All tasks retuned successfully, and teh schema was updated successfully
    RunSuccess = 4,
    /// something fucked up
    RunFailure = 5,
}

impl MigrationJobStatus {
    pub fn from_int(i: u64) -> Self {
        match i {
            0 => Self::WaitingDryRun,
            1 => Self::DryRunSuccess,
            2 => Self::DryRunFailure,
            3 => Self::WaitingRun,
            4 => Self::RunSuccess,
            5 => Self::RunFailure,
            _ => panic!(),
        }
    }

    /// Returns `true` if the migration job status is [`WaitingRun`].
    ///
    /// [`WaitingRun`]: MigrationJobStatus::WaitingRun
    #[must_use]
    pub fn is_waiting_run(&self) -> bool {
        matches!(self, Self::WaitingRun)
    }
}
