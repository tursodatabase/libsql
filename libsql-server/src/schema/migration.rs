use itertools::Itertools;
use once_cell::sync::Lazy;
use rusqlite::Savepoint;

use crate::connection::program::{Program, Vm};
use crate::query_result_builder::{IgnoreResult, QueryBuilderConfig, QueryResultBuilder};

use super::result_builder::SchemaMigrationResultBuilder;
use super::status::MigrationTask;
use super::{Error, MigrationTaskStatus};

pub fn setup_migration_table(conn: &mut rusqlite::Connection) -> Result<(), Error> {
    static TASKS_TABLE_QUERY: Lazy<String> = Lazy::new(|| {
        format!(
            "CREATE TABLE IF NOT EXISTS __libsql_migration_tasks (
                job_id INTEGER PRIMARY KEY,
                status INTEGER,
                migration TEXT NOT NULL,
                error TEXT,
                finished BOOLEAN GENERATED ALWAYS AS ({})
            )",
            MigrationTaskStatus::finished_states()
                .iter()
                .map(|s| format!("status = {}", *s as u64))
                .join(" OR ")
        )
    });

    let tx = conn.transaction()?;
    let schema_version =
        tx.pragma_query_value(None, "schema_version", |row| Ok(row.get::<_, u64>(0)?))?;
    tx.execute(&*TASKS_TABLE_QUERY, ())?;
    // We have to make sure schema_version is not changed in case this db is a fork of shared schema
    // or a namespace that links to shared schema.
    // It is ok to do this because we're the only connection to that database at this point.
    tx.pragma_update(None, "schema_version", schema_version)?;
    tx.commit()?;
    Ok(())
}

pub fn has_pending_migration_task(conn: &rusqlite::Connection) -> Result<bool, Error> {
    Ok(conn.query_row(
        "SELECT COUNT(1) FROM __libsql_migration_tasks WHERE finished = false",
        (),
        |row| Ok(row.get::<_, usize>(0)? != 0),
    )?)
}

pub fn enqueue_migration_task(
    conn: &rusqlite::Connection,
    task: &MigrationTask,
    migration: &Program,
) -> Result<(), Error> {
    let migration = serde_json::to_string(migration).unwrap();
    conn.execute(
        "INSERT OR IGNORE INTO __libsql_migration_tasks (job_id, status, migration) VALUES (?, ?, ?)",
        (task.job_id(), *task.status() as u64, &migration),
    )?;

    Ok(())
}

pub fn abort_migration_task(
    conn: &rusqlite::Connection,
    task: &MigrationTask,
) -> Result<(), Error> {
    // there is a `NOT NULL` constraint on migration, but if we are aborting a task that wasn't
    // already enqueued, we need a placeholder. It's ok because we are never gonna try to run a
    // failed task migration.
    conn.execute("INSERT OR REPLACE INTO __libsql_migration_tasks (job_id, status, error, migration) VALUES (?, ?, ?, ?)",
    (task.job_id, MigrationTaskStatus::Failure as u64, "aborted", "aborted"))?;

    Ok(())
}

/// set the task status to `Run` if its current state is `DryRunSuccess`
pub fn step_migration_task_run(conn: &rusqlite::Connection, job_id: i64) -> Result<(), Error> {
    conn.execute(
        "
            UPDATE __libsql_migration_tasks
            SET status = ?
            WHERE job_id = ? AND status = ?
            ",
        (
            MigrationTaskStatus::Run as u64,
            job_id,
            MigrationTaskStatus::DryRunSuccess as u64,
        ),
    )?;

    Ok(())
}

fn get_task_infos(
    conn: &rusqlite::Connection,
    job_id: i64,
) -> Result<(MigrationTaskStatus, Option<Program>, Option<String>), Error> {
    Ok(conn.query_row(
        "SELECT status, migration, error FROM __libsql_migration_tasks WHERE job_id = ?",
        [job_id],
        |row| {
            let status = MigrationTaskStatus::from_int(row.get::<_, u64>(0)?);
            let (migration, error) = match status {
                MigrationTaskStatus::Enqueued | MigrationTaskStatus::Run => {
                    let migration: Program =
                        serde_json::from_str(row.get_ref(1)?.as_str()?).unwrap();
                    (Some(migration), None)
                }
                MigrationTaskStatus::DryRunSuccess | MigrationTaskStatus::Success => (None, None),
                MigrationTaskStatus::DryRunFailure | MigrationTaskStatus::Failure => {
                    let error = row.get::<_, Option<String>>(2)?;
                    (None, error)
                }
            };

            Ok((status, migration, error))
        },
    )?)
}

pub(super) fn step_task(
    txn: &mut rusqlite::Transaction,
    job_id: i64,
) -> Result<(MigrationTaskStatus, Option<String>), Error> {
    let (current_state, migration, error) = get_task_infos(txn, job_id)?;

    match current_state {
        MigrationTaskStatus::DryRunSuccess | MigrationTaskStatus::DryRunFailure => {
            Ok((current_state, error))
        }
        MigrationTaskStatus::Run | MigrationTaskStatus::Enqueued => {
            let (ret, new_status) = perform_migration(
                txn,
                migration.as_ref().unwrap(),
                current_state.is_enqueued(),
                IgnoreResult,
                &QueryBuilderConfig::default(),
            );
            let error = ret.err().map(|e| e.to_string());
            update_db_task_status(txn, job_id, new_status, error.as_deref())?;

            Ok((new_status, error))
        }
        // final state, nothing to do but report
        MigrationTaskStatus::Success | MigrationTaskStatus::Failure => Ok((current_state, error)),
    }
}

pub fn perform_migration<B: QueryResultBuilder>(
    txn: &mut rusqlite::Transaction,
    migration: &Program,
    dry_run: bool,
    builder: B,
    config: &QueryBuilderConfig,
) -> (Result<B, Error>, MigrationTaskStatus) {
    // todo error handling is sketchy, improve
    let builder = SchemaMigrationResultBuilder::new(builder);
    let mut savepoint = txn.savepoint().unwrap();
    match try_perform_migration(&mut savepoint, migration, builder, config) {
        Ok(builder) if builder.is_success() => {
            let status = if dry_run {
                savepoint.rollback().unwrap();
                drop(savepoint);
                MigrationTaskStatus::DryRunSuccess
            } else {
                savepoint.commit().unwrap();
                MigrationTaskStatus::Success
            };
            (Ok(builder.into_inner()), status)
        }
        Ok(builder) => {
            assert!(!builder.is_success());
            savepoint.rollback().unwrap();
            drop(savepoint);
            let status = if dry_run {
                MigrationTaskStatus::DryRunFailure
            } else {
                MigrationTaskStatus::Failure
            };
            let (step, error) = builder.into_error();
            (Err(Error::MigrationError(step, error)), status)
        }
        Err(e) => {
            savepoint.rollback().unwrap();
            drop(savepoint);
            let status = if dry_run {
                MigrationTaskStatus::DryRunFailure
            } else {
                MigrationTaskStatus::Failure
            };
            (Err(e), status)
        }
    }
}

pub(super) fn update_db_task_status(
    conn: &rusqlite::Connection,
    job_id: i64,
    status: MigrationTaskStatus,
    error: Option<&str>,
) -> Result<(), Error> {
    conn.execute(
        "UPDATE __libsql_migration_tasks SET status = ?, error = ? WHERE job_id = ?",
        (status as u64, error, job_id),
    )?;

    Ok(())
}

fn try_perform_migration<B: QueryResultBuilder>(
    savepoint: &mut Savepoint,
    migration: &Program,
    mut builder: B,
    config: &QueryBuilderConfig,
) -> Result<B, Error> {
    builder.init(config).unwrap();
    let mut vm = Vm::new(
        builder, // todo use proper builder
        migration,
        |_| (false, None),
        |_, _, _| (),
    );

    while !vm.finished() {
        vm.step(savepoint).unwrap(); // return migration error
    }

    vm.builder().finish(None, true).unwrap();

    Ok(vm.into_builder())
}

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;
    use libsql_sys::wal::Sqlite3WalManager;
    use tempfile::tempdir;

    use crate::connection::libsql::open_conn_active_checkpoint;
    use crate::namespace::NamespaceName;

    use super::*;

    #[test]
    fn already_performed_task_is_not_reexecuted() {
        let tmp = tempdir().unwrap();

        let mut conn =
            open_conn_active_checkpoint(tmp.path(), Sqlite3WalManager::default(), None, 1000, None)
                .unwrap();
        setup_migration_table(&mut conn).unwrap();

        let task = MigrationTask {
            namespace: NamespaceName::default(),
            status: MigrationTaskStatus::Success,
            job_id: 1,
            task_id: 1,
        };
        enqueue_migration_task(&conn, &task, &Program::seq(&["create table test (x)"])).unwrap();
        let mut txn = conn.transaction().unwrap();
        let (status, error) = step_task(&mut txn, 1).unwrap();
        txn.commit().unwrap();

        assert!(error.is_none());
        assert_eq!(status, MigrationTaskStatus::Success);

        // The migration should not have been performed, since we declared that it was already
        // successfully executed.
        let schema = conn
            .prepare("select * from sqlite_schema")
            .unwrap()
            .query_map((), |r| Ok(format!("{r:?}")))
            .unwrap()
            .collect::<Vec<_>>();
        assert_debug_snapshot!(schema);
    }

    #[test]
    fn ignore_task_enqueue_if_already_exists() {
        let tmp = tempdir().unwrap();

        let mut conn =
            open_conn_active_checkpoint(tmp.path(), Sqlite3WalManager::default(), None, 1000, None)
                .unwrap();
        setup_migration_table(&mut conn).unwrap();

        let task = MigrationTask {
            namespace: NamespaceName::default(),
            status: MigrationTaskStatus::Success,
            job_id: 1,
            task_id: 1,
        };
        enqueue_migration_task(&conn, &task, &Program::seq(&["create table test (x)"])).unwrap();

        let task = MigrationTask {
            namespace: NamespaceName::default(),
            status: MigrationTaskStatus::Enqueued,
            job_id: 1,
            task_id: 1,
        };
        enqueue_migration_task(&conn, &task, &Program::seq(&["create table test (x)"])).unwrap();
        let (status, _, _) = get_task_infos(&conn, 1).unwrap();
        assert_eq!(status, MigrationTaskStatus::Success);
    }
}
