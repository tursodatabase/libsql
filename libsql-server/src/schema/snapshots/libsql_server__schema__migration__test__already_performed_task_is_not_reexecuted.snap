---
source: libsql-server/src/schema/migration.rs
expression: schema
---
[
    Ok(
        "{Ok(\"type\"): (Text, \"table\"), Ok(\"name\"): (Text, \"sqlite3_libsql_tasks\"), Ok(\"tbl_name\"): (Text, \"sqlite3_libsql_tasks\"), Ok(\"rootpage\"): (Integer, 2), Ok(\"sql\"): (Text, \"CREATE TABLE sqlite3_libsql_tasks (\\n                job_id INTEGER PRIMARY KEY,\\n                status INTEGER,\\n                migration TEXT NOT NULL,\\n                error TEXT,\\n                finished BOOLEAN GENERATED ALWAYS AS (status = 4 OR status = 5)\\n            )\")}",
    ),
]
